### PR TITLE
Add uploading snapshot to cloud feature

### PR DESCRIPTION
**Pull request by xxwjj**
**What this PR does / why we need it**:
Add uploading snapshot to cloud feature

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: backup/driver.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+type ctorFun func() (BackupDriver, error)
+
+var ctorFunMap = map[string]ctorFun{}
+
+func NewBackup(backupDriverName string) (BackupDriver, error) {
+	fun, exist := ctorFunMap[backupDriverName]
+	if !exist {
+		return nil, fmt.Errorf("specified backup driver is not exist.")
```
Comment (xing-yang) [inline backup/driver.go]: s/is not exist/does not exist

File: backup/multicloud/client.go [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+	}
+
+	fullUrl := base.ResolveReference(u)
+	b, _, err := c.doRequest(method, fullUrl.String(), in, cb)
+	if err != nil {
+		return err
+	}
+
+	if out != nil {
+		log.V(8).Infof("Response:\n%s\n", string(b))
```
Comment (xing-yang) [inline backup/multicloud/client.go]: V(5) should be good for debug msg.

File: backup/multicloud/client.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+}
+
+func (c *Client) UploadPart(bucketName, objectKey string, partNum int64, uploadId string, data []byte, size int64) (*UploadPartResult, error) {
+	log.Infof("upload part buf size:%d", len(data))
+	p := path.Join("s3", bucketName, objectKey)
+	p += fmt.Sprintf("?partNumber=%d&uploadId=%s", partNum, uploadId)
+	out := &UploadPartResult{}
+	reqSettingCB := func(req *httplib.BeegoHTTPRequest) error {
+		req.Header("Content-Length", strconv.FormatInt(size, 10))
+		req.SetTimeout(30*time.Minute, 30*time.Minute)
```
Comment (xing-yang) [inline backup/multicloud/client.go]: Add a TODO to make the timeout value configurable.

File: backup/multicloud/client.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+	p += fmt.Sprintf("?uploadId=%s", uploadId)
+	out := &CompleteMultipartUploadResult{}
+	if err := c.request("PUT", p, input, nil, nil); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+func (c *Client) AbortMultipartUpload(bucketName, objectKey string) error {
+	//p := path.Join("s3", "AbortMultipartUpload", bucketName, objectKey)
```
Comment (xing-yang) [inline backup/multicloud/client.go]: Why are these commented out?  If this does not work yet, add a TODO to implement it later and delete unused code.

File: backup/multicloud/driver.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+	"github.com/opensds/opensds/backup"
+	"gopkg.in/yaml.v2"
+	"io"
+	"io/ioutil"
+	"os"
+)
+
+const (
+	ConfFile        = "/etc/opensds/driver/multi-cloud.yaml"
+	UploadChunkSize = 1024 * 1024 * 5
```
Comment (xing-yang) [inline backup/multicloud/driver.go]: Add a TODO to make this configurable.

File: contrib/drivers/huawei/dorado/dorado.go [hunk 6]
```diff
@@ -723,6 +723,14 @@ func (d *Driver) clearHostRelatedResource(lunGrpId, viewId, hostId, hostGrpId st
 	return nil
 }
 
+func (d *Driver) InitializeSnapshotConnection(opt *pb.CreateSnapshotAttachmentOpts) (*model.ConnectionInfo, error) {
+	return nil, &model.NotImplementError{S: "Method InitializeSnapshotConnection has not implement yet."}
```
Comment (xing-yang) [inline contrib/drivers/huawei/dorado/dorado.go]: s/has not implement/has not been implemented

File: contrib/drivers/huawei/dorado/dorado.go [hunk 7]
```diff
@@ -723,6 +723,14 @@ func (d *Driver) clearHostRelatedResource(lunGrpId, viewId, hostId, hostGrpId st
...
+}
+
+func (d *Driver) TerminateSnapshotConnection(opt *pb.DeleteSnapshotAttachmentOpts) error {
+	return &model.NotImplementError{S: "Method TerminateSnapshotConnection has not implement yet."}
```
Comment (xing-yang) [inline contrib/drivers/huawei/dorado/dorado.go]: s/has not implement/has not been implemented

File: contrib/drivers/lvm/lvm.go [hunk 8]
```diff
@@ -23,13 +23,15 @@ import (
...
 	log "github.com/golang/glog"
+	"github.com/opensds/opensds/backup"
 	"github.com/opensds/opensds/contrib/drivers/lvm/targets"
 	. "github.com/opensds/opensds/contrib/drivers/utils/config"
 	pb "github.com/opensds/opensds/pkg/dock/proto"
 	"github.com/opensds/opensds/pkg/model"
 	"github.com/opensds/opensds/pkg/utils"
 	"github.com/opensds/opensds/pkg/utils/config"
 	"github.com/satori/go.uuid"
+	"os"
```
Comment (xing-yang) [inline contrib/drivers/lvm/lvm.go]: Have you run go fmt?  I thought "os" is usually in the top import block.

File: contrib/drivers/lvm/lvm.go [hunk 9]
```diff
@@ -545,6 +626,14 @@ func (d *Driver) ListPools() ([]*model.StoragePoolSpec, error) {
 	return pols, nil
 }
 
+func (d *Driver) InitializeSnapshotConnection(opt *pb.CreateSnapshotAttachmentOpts) (*model.ConnectionInfo, error) {
+	return nil, &model.NotImplementError{S: "Method InitializeSnapshotConnection has not implement yet."}
+}
+
+func (d *Driver) TerminateSnapshotConnection(opt *pb.DeleteSnapshotAttachmentOpts) error {
+	return &model.NotImplementError{S: "Method TerminateSnapshotConnection has not implement yet."}
```
Comment (xing-yang) [inline contrib/drivers/lvm/lvm.go]: InitializeSnapshotConnection and TerminateSnapshotConnection in LVM driver should be implemented.  It should just use the local path without the real attaching/detaching.

File: contrib/drivers/ceph/ceph.go [hunk 10]
```diff
@@ -601,6 +601,14 @@ func (d *Driver) ListPools() ([]*model.StoragePoolSpec, error) {
 	return pols, nil
 }
 
+func (d *Driver) InitializeSnapshotConnection(opt *pb.CreateSnapshotAttachmentOpts) (*model.ConnectionInfo, error) {
+	return nil, &model.NotImplementError{S: "Method InitializeSnapshotConnection has not implement yet."}
```
Comment (xing-yang) [inline contrib/drivers/ceph/ceph.go]: s/has not implement/has not been implemented

File: contrib/drivers/ceph/ceph.go [hunk 11]
```diff
@@ -601,6 +601,14 @@ func (d *Driver) ListPools() ([]*model.StoragePoolSpec, error) {
...
+}
+
+func (d *Driver) TerminateSnapshotConnection(opt *pb.DeleteSnapshotAttachmentOpts) error {
+	return &model.NotImplementError{S: "Method TerminateSnapshotConnection has not implement yet."}
```
Comment (xing-yang) [inline contrib/drivers/ceph/ceph.go]: s/has not implement/has not been implemented

File: contrib/drivers/openstack/cinder/cinder.go [hunk 12]
```diff
@@ -462,6 +462,14 @@ func (d *Driver) ListPools() ([]*model.StoragePoolSpec, error) {
 	return typePol, nil
 }
 
+func (d *Driver) InitializeSnapshotConnection(opt *pb.CreateSnapshotAttachmentOpts) (*model.ConnectionInfo, error) {
+	return nil, &model.NotImplementError{S: "Method InitializeSnapshotConnection has not implement yet."}
```
Comment (xing-yang) [inline contrib/drivers/openstack/cinder/cinder.go]: s/has not implement/has not been implemented

File: contrib/drivers/openstack/cinder/cinder.go [hunk 13]
```diff
@@ -462,6 +462,14 @@ func (d *Driver) ListPools() ([]*model.StoragePoolSpec, error) {
...
+}
+
+func (d *Driver) TerminateSnapshotConnection(opt *pb.DeleteSnapshotAttachmentOpts) error {
+	return &model.NotImplementError{S: "Method TerminateSnapshotConnection has not implement yet."}
```
Comment (xing-yang) [inline contrib/drivers/openstack/cinder/cinder.go]: s/has not implement/has not been implemented

File: testutils/driver/sample.go [hunk 14]
```diff
@@ -102,6 +102,14 @@ func (*Driver) ListPools() ([]*model.StoragePoolSpec, error) {
 	return pols, nil
 }
 
+func (d *Driver) InitializeSnapshotConnection(opt *pb.CreateSnapshotAttachmentOpts) (*model.ConnectionInfo, error) {
+	return nil, &model.NotImplementError{S: "Method InitializeSnapshotConnection has not implement yet."}
```
Comment (xing-yang) [inline testutils/driver/sample.go]: s/has not implement/has not been implemented

File: testutils/driver/sample.go [hunk 15]
```diff
@@ -102,6 +102,14 @@ func (*Driver) ListPools() ([]*model.StoragePoolSpec, error) {
...
+}
+
+func (d *Driver) TerminateSnapshotConnection(opt *pb.DeleteSnapshotAttachmentOpts) error {
+	return &model.NotImplementError{S: "Method TerminateSnapshotConnection has not implement yet."}
```
Comment (xing-yang) [inline testutils/driver/sample.go]: s/has not implement/has not been implemented

File: backup/multicloud/client.go (hunk 4 above)
Comment (xxwjj) [reply backup/multicloud/client.go]: done

File: pkg/dock/proto/dock.proto [hunk 16]
```diff
@@ -236,6 +236,49 @@ message DeleteAttachmentOpts {
     string AccessProtocol = 7;
 }
 
+
```
Comment (leonwanghui) [inline pkg/dock/proto/dock.proto]: There should only be one blank line here

**Pull request merged**